### PR TITLE
Deleting all Set-Cookie headers in the response

### PR DESCRIPTION
**Pull request by Doerge**
All Set-Cookie headers are now deleted on the response, before the zappa cookie is set.

Added comments to make it (hopefully) make it clearer what is going on.

### CODE REVIEW

**Comment by bbangert** [comment]
LGTM

**Comment by Miserlou** [comment]
Looking good, but do I need to merge before we have the expiry working? 

Also - since this is critical and complicated code, a test would be GREATLY appreciated..

**Comment by Doerge** [comment]
There were some  ugly stuff going on in the middleware. Rewrote most of the logic and added 700 lines of tests.

I still have trouble reproducing bugs from the production environment, though :8ball:

File: tests/tests_middleware.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        # Assert that the simple_app, received the decoded cookies
+        excpected = {'foo': '123', 'bar': '456', 'baz': '789'}
+        received = ''.join(resp)
+        # Split the response on ';', then on '=', then convert to dict for
+        # comparison
+        received = dict([cookie.split('=') for cookie in received.split(';')])
+        self.assertDictEqual(received, excpected)
+
+    def test_wsgi_middleware_cookieoverwrite(self):
+        """ This method do:
```
Comment (collingreen) [inline tests/tests_middleware.py]: nitpick:  remove 'do'

File: tests/tests_middleware.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        # Call with empty WSGI Environment
+        resp = app(dict(), self._start_response)
+
+        # Ensure the encoded zappa cookie is set
+        self.assertEqual(self.headers[0][0], 'Set-Cookie')
+        zappa_cookie = self.headers[0][1]
+        self.assertTrue(zappa_cookie.startswith('zappa='))
+
+        # Reads the hopefully decoded cookies
+        def simple_app(environ, start_response):
```
Comment (collingreen) [inline tests/tests_middleware.py]: nitpick - redefined simple_app in this test

**Pull request merged**